runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,110 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
+                  // 1.) The object's type signature is unavailable/unknown
+                  // 2.) The signature names java/lang/Object
+                  // 3.) The signature names some interface type, wich is a problem because interface types are
+                  //     not checked by the verifier
+                  // 4.) The signature belongs to a parameter of the calling method, which is a problem because
+                  //     parameters can share slots with variables of other, potentially incompatible types
+                  bool objTypeUnknown;
+                  TR_OpaqueClasBlock *objClass = NULL;
+
+                  if (objTypeSig == NULL)
+                     {
+                     objTypeUnknown = true;
+                     }
+                  else
+                     {
+                     objClass = fej9->getClassFromSignature(objTypeSig, length, dest->getSymbolReference()->getOwningMethod(comp));
+                     bool isParameter = dest->getSymbolReference()->getSymbol()->isParm();
+
+                     objTypeUnknown = objClass == NULL || objClass == comp->getObjectClassPointer() ||
+                                      TR::Compiler->cls.isInterfaceClass(comp, objClass) || isParameter;
+                     }
+
+                  //need to generate code to load dataAddr in cases (2) and (3) (unless the object is known to be NULL)
+                  mayNeedDataAddrLoad = !dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');
+
+                  //pass in dest and destOffset separately in case (3) only
+                  separateDestAndOffset = mayNeedDataAddrLoad && objTypeUnknown;
+                  }
+               else
+                  {
+                  //if for some reason we cannot get the symbol reference (and thus, the type signature)
+                  //of dest, we'll need to take the conservative approach and pass dest and destOffset to
+                  //setmemoryEvaluator() separately
+                  separateDestAndOffset = true;
+                  mayNeedDataAddrLoad = true;
+                  }
+            }
+         #endif /* J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION */
+
+            TR::Node *copyMemNode;
+
+            //setting up IL trees to be passed along to setmemoryEvaluator()
+            if (separateDestAndOffset) // CASE (3): dest and destOffset passed in separately
+               {
+               //four child setup
+               copyMemNode = TR::Node::createWithSymRef(TR::arrayset, 4, 4, dest, destOffset, len, byteValue, node->getSymbolReference());
+               }
+            else // CASE (1) and (2): dest += destoffset, then pass in to evaluator
+               {
+            #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+
+               if (mayNeedDataAddrLoad) // CASE (2) only
+                  {
+                  //load dataAddr and use as object base address (previously dest)
```

Comment:
i still don't think this is correct, as I mentioned previously:  let's think about a simple case ... it is an array type but runtime passes in null -- look back to the original condition:    `!dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');` ... so, `!dest->isNull()` is true (since we don't know if it is null at compile-time) ... below code trying to load dataAddr will crash.

Reply:
Good point, thank you for catching that. I think this be addressed by modifying `separateDestAndOffset` rather than `mayNeedDataAddrLoad` though:
```
//need to generate code to load dataAddr in cases (2) and (3) (unless the object is known to be NULL)
mayNeedDataAddrLoad = !dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');

//pass in dest and destOffset separately in case (3) OR in case (2) if we don't know for sure that the object is non-NULL
separateDestAndOffset = mayNeedDataAddrLoad && (objTypeUnknown || !dest->isNonNull());
```
This will ensure that any runtime NULL, even from an object that is known at compile time to be an array, is caught before attempting to load dataAddr (since `setmemoryEvaluator()` generates a null test just before the array check).

Reply:
mayNeedDataAddrLoad appeared to be a (anti) sub-condition of separateDestAndOffset (i.e. the former is never used if the latter is true). but, here, you are setting them up as if the latter is dependent on the former being true. def-use logic is pretty odd. have you thought through them covering all possible cases?

Reply:
Initially the idea there was that `mayNeedDataAddrLoad == true` encapsulated either case (2) or case (3), and that the value of `separateDestAndOffset` would differentiate those two cases from each other (i.e.: `separateDestAndOffset == true` for case (3), `separateDestAndOffset == false` for case (2)). However, with additional NULL/non-NULL conditions, that correlation is not quite as clean cut.

If it would make the code easier to understand, I can modify the logic behind those conditions so that they're independent from each other?

Reply:
if you can, why not?